door-open-service/node-backend/src/DoorBackend.js

```diff
@@ -0,0 +1,49 @@
+import _ from "underscore";
+
+class DoorBackend {
+
+  constructor(config) {
+    this.doorConfigs = config.doorConfigs;
+    this.locks = _.mapObject(config.doorConfigs, doorConfig =>
+      (doorConfig.lockTime ? { lockTime: doorConfig.lockTime } : undefined)
+    );
+    this.openers = [];
+  }
+
+  addOpener(opener) {
+    this.openers.push(opener);
+  }
+
+  open(user, doorId) {
+    // check if the user was properly authenticated (handled by DoorSlackBot)
+    if (!user) return { result: DoorBackend.OPEN_RESULT.UNAUTHORIZED };
+
+    // check if the door is locked by the virtue of a recent opening by another person
+    const lock = this.locks[doorId];
+    if (lock && lock.lastUser !== user.email) {
+      const now = new Date().getTime();
+      if (now - lock.lastTime < lock.lockTime) {
+        return { result: DoorBackend.OPEN_RESULT.LOCKED, user: lock.lastUser };
+      }
+      lock.lastUser = user.email;
+      lock.lastTime = now;
```

Comment:
Shouldn't we always update the `lastTime`, even if it's the same user issuing the request? I think it can help in situations where the opener doesn't work for some reason and the user needs to retry. As it is, we won't update the lock time and other user can try and open it outside the "expected" lock time.

Reply:
Roger. Changed in d485a2a.